Support Python 3

oliverlee opened the pull request:
Addresses https://github.com/pydy/pydy/issues/38
I got 2 of the examples (mass_spring_damper, three_link_conical_pendulum) to run with Python 3.4.2.

- [x] There are no merge conflicts.
- [x] If there is a related issue, a reference to that issue is in the
  commit message.
- [x] Unit tests have been added for the new feature.
- [x] The PR passes tests both locally (run `nosetests`) and on Travis CI.
- [x] All public methods and classes have docstrings. (We use the [numpydoc
  format](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt).)
- [x] An explanation has been added to the online documentation. (`docs`
  directory)
- [x] The code follows PEP8 guidelines. (use a linter, e.g.
  [pylint](http://www.pylint.org), to check your code)
- [x] The new feature is documented in the [Release
  Notes](https://github.com/pydy/pydy#release-notes).
- [x] The code is backwards compatible. (All public methods/classes must
  follow deprecation cycles.)
- [x] All reviewer comments have been addressed.

oliverlee commented:
Also I don't know if Python3' or 'Python 3' is more correct.

moorepants commented:
This is great. I'm sure it will take some testing tweaking. We'll see if it passes all of the Python 2.7 tests.

I'd also like to get Travis running Python 3 but we may need to switch to using miniconda as the testing backend so that we can easily pull in the dependencies.

Do you think we should support Python 3.3+? Or just Python 3.4+?

@pydy/pydy-developers This needs review. All help is welcome.

oliverlee commented:
3.4

oliverlee commented:
I also looked through all the sources after using 2to3 and it seems okay to me.

moorepants commented:
I just got a bunch of test failures locally testing with Python 3.4:

```
======================================================================
ERROR: pydy.viz.tests.test_scene.TestScene.test_create_static_html
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/viz/tests/test_scene.py", line 247, in test_create_static_html
    outfile_prefix="test")
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 364, in generate_visualization_json_system
    system.constants.values(), **kwargs)
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 208, in generate_visualization_json
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 312, in generate_simulation_dict
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/visualization_frame.py", line 290, in evaluate_transformation_matrix
    new[i, :, :] = self._numeric_transform(*args)
TypeError: <lambda>() missing 3 required positional arguments: '_Dummy_6742', '_Dummy_6743', and '_Dummy_6744'

======================================================================
ERROR: pydy.viz.tests.test_scene.TestScene.test_custom_camera
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/viz/tests/test_scene.py", line 230, in test_custom_camera
    self.sys.constants.values())
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 312, in generate_simulation_dict
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/visualization_frame.py", line 290, in evaluate_transformation_matrix
    new[i, :, :] = self._numeric_transform(*args)
TypeError: <lambda>() missing 3 required positional arguments: '_Dummy_6771', '_Dummy_6772', and '_Dummy_6773'

======================================================================
ERROR: pydy.viz.tests.test_scene.TestScene.test_generate_scene_dict
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/viz/tests/test_scene.py", line 156, in test_generate_scene_dict
    self.sys.constants.values())
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 312, in generate_simulation_dict
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/visualization_frame.py", line 290, in evaluate_transformation_matrix
    new[i, :, :] = self._numeric_transform(*args)
TypeError: <lambda>() missing 3 required positional arguments: '_Dummy_6787', '_Dummy_6788', and '_Dummy_6789'

======================================================================
ERROR: pydy.viz.tests.test_scene.TestScene.test_generate_simulation_dict
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/viz/tests/test_scene.py", line 111, in test_generate_simulation_dict
    self.sys.constants.values())
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 312, in generate_simulation_dict
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/visualization_frame.py", line 290, in evaluate_transformation_matrix
    new[i, :, :] = self._numeric_transform(*args)
TypeError: <lambda>() missing 3 required positional arguments: '_Dummy_6803', '_Dummy_6804', and '_Dummy_6805'

======================================================================
ERROR: pydy.viz.tests.test_scene.TestScene.test_generate_visualization_json_system
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/viz/tests/test_scene.py", line 263, in test_generate_visualization_json_system
    scene.generate_visualization_json_system(self.sys)
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 364, in generate_visualization_json_system
    system.constants.values(), **kwargs)
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 208, in generate_visualization_json
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/scene.py", line 312, in generate_simulation_dict
    constant_values)
  File "/home/moorepants/src/pydy/pydy/viz/visualization_frame.py", line 290, in evaluate_transformation_matrix
    new[i, :, :] = self._numeric_transform(*args)
TypeError: <lambda>() missing 3 required positional arguments: '_Dummy_6819', '_Dummy_6820', and '_Dummy_6821'

======================================================================
ERROR: pydy.tests.test_system.TestSystem.test_constants
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/tests/test_system.py", line 110, in test_constants
    testing.assert_allclose(sys.constants.values(), constants.values())
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 1297, in assert_allclose
    verbose=verbose, header=header)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 648, in assert_array_compare
    val = comparison(x, y)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 1292, in compare
    return np.allclose(x, y, rtol=rtol, atol=atol)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/core/numeric.py", line 2219, in allclose
    xinf = isinf(x)
TypeError: ufunc 'isinf' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

======================================================================
ERROR: pydy.tests.test_system.TestSystem.test_initial_conditions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/tests/test_system.py", line 277, in test_initial_conditions
    testing.assert_allclose(sys.initial_conditions.values(), ic.values())
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 1297, in assert_allclose
    verbose=verbose, header=header)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 648, in assert_array_compare
    val = comparison(x, y)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/testing/utils.py", line 1292, in compare
    return np.allclose(x, y, rtol=rtol, atol=atol)
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/numpy/core/numeric.py", line 2219, in allclose
    xinf = isinf(x)
TypeError: ufunc 'isinf' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_doprint
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/codegen/tests/test_c_code.py", line 319, in test_doprint
    assert header == expected_header
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_code_blocks
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/codegen/tests/test_c_code.py", line 212, in test_generate_code_blocks
    assert v == expected[k]
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_comma_lists
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/codegen/tests/test_c_code.py", line 108, in test_generate_comma_lists
    assert self.generator.comma_lists() == expected
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_cse
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/codegen/tests/test_c_code.py", line 86, in test_generate_cse
    assert self.generator.subexprs == expected_subexprs
AssertionError

======================================================================
FAIL: pydy.tests.test_models.test_multi_mass_spring_damper_double
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/tests/test_models.py", line 71, in test_multi_mass_spring_damper_double
    assert sys.specifieds_symbols == [f0, f1]
AssertionError

======================================================================
FAIL: pydy.tests.test_system.TestSystem.test_init
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/tests/test_system.py", line 63, in test_init
    assert sys.specifieds.keys() == [dynamicsymbols('f0')]
AssertionError

======================================================================
FAIL: pydy.tests.test_system.TestSystem.test_specifieds
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/moorepants/anaconda/envs/pydy-dev-py3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/moorepants/src/pydy/pydy/tests/test_system.py", line 150, in test_specifieds
    assert sys.specifieds.keys() == [dynamicsymbols('f0')]
AssertionError

----------------------------------------------------------------------
Ran 67 tests in 123.297s

FAILED (errors=7, failures=7)
(pydy-dev-py3)moorepants@moorepants-2170p:pydy(pr-229)$ conda list
# packages in environment at /home/moorepants/anaconda/envs/pydy-dev-py3:
#
alabaster                 0.7.3                    py34_0  
babel                     1.3                      py34_0  
certifi                   14.05.14                 py34_0  
cython                    0.22                     py34_0  
dateutil                  2.1                      py34_2  
docutils                  0.12                     py34_0  
fontconfig                2.11.1                        4  
freetype                  2.5.2                         2  
ipython                   3.1.0                    py34_1  
ipython-notebook          3.1.0                    py34_1  
jinja2                    2.7.3                    py34_1  
jsonschema                2.4.0                    py34_0  
libpng                    1.6.17                        0  
libsodium                 0.4.5                         0  
libxml2                   2.9.0                         0  
markupsafe                0.23                     py34_0  
matplotlib                1.4.3                np19py34_2  
mistune                   0.5.1                    py34_1  
modernize                 0.4                       <pip>
mpmath                    0.19                     py34_0  
nose                      1.3.7                    py34_0  
numpy                     1.9.2                    py34_0  
openssl                   1.0.1k                        1  
pip                       7.0.3                    py34_0  
ptyprocess                0.4                      py34_0  
pygments                  2.0.2                    py34_0  
pyparsing                 2.0.3                    py34_0  
pyqt                      4.11.3                   py34_1  
python                    3.4.3                         1  
python-dateutil           2.4.2                    py34_0  
pytz                      2015.4                   py34_0  
pyzmq                     14.6.0                   py34_0  
qt                        4.8.6                         3  
readline                  6.2                           2  
scipy                     0.15.1               np19py34_0  
setuptools                17.0                     py34_0  
sip                       4.16.5                   py34_0  
six                       1.9.0                    py34_0  
snowballstemmer           1.2.0                    py34_0  
sphinx                    1.3.1                    py34_0  
sphinx-rtd-theme          0.1.7                     <pip>
sphinx_rtd_theme          0.1.7                    py34_0  
sqlite                    3.8.4.1                       1  
sympy (/home/moorepants/src/sympy) 0.7.7.dev0                <pip>
system                    5.8                           2  
terminado                 0.5                      py34_0  
theano                    0.7.0                np19py34_0  
tk                        8.5.18                        0  
tornado                   4.2                      py34_0  
xz                        5.0.5                         0  
zeromq                    4.0.5                         0  
zlib                      1.2.8                         0  
```

oliverlee commented:
Oh I didn't run the tests correctly.

moorepants commented:
I do `nosetests -v` from the top level directory.

oliverlee commented:
$ nosetests -v
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 651, in _build_master
    ws.require(__requires__)
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 952, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 844, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (nose 1.3.7 (/usr/local/lib/python3.4/dist-packages), Requirement.parse('nose==1.3.4'))

moorepants commented:
I can't work on this today, but I'll help setup Travis for Python 2 and 3 testing so we can merge that in with this PR (unless you want to do it). I think we should switch Travis to using miniconda, bump the minimum dep versions for numpy/scipy/etc up to the whatever the oldest versions are in the anaconda repos, and then test on 2.7, maybe 3.3 (sympy supports 3.3), and 3.4.

moorepants commented:
Do you have `nose` installed?

oliverlee commented:
oliver@arcturus:~$ nose
No command 'nose' found, did you mean:

Probably don't have time to setup Travis.

moorepants commented:
The command is `nosetests` and the package name is `nose`, i.e. `pip install nose`.

oliverlee commented:
Oh yeah, that was already installed. I thought you meant some other tool called ```nose```.

oliverlee commented:
Nvm I figured it out.

oliverlee commented:
Some tests fail on my computer:
 ```
======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_doprint
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/oliver/repos/math/pydy/pydy/codegen/tests/test_c_code.py", line 319, in test_doprint
    assert header == expected_header
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_code_blocks
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/oliver/repos/math/pydy/pydy/codegen/tests/test_c_code.py", line 212, in test_generate_code_blocks
    assert v == expected[k]
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_comma_lists
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/oliver/repos/math/pydy/pydy/codegen/tests/test_c_code.py", line 108, in test_generate_comma_lists
    assert self.generator.comma_lists() == expected
AssertionError

======================================================================
FAIL: pydy.codegen.tests.test_c_code.TestCMatrixGenerator.test_generate_cse
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/oliver/repos/math/pydy/pydy/codegen/tests/test_c_code.py", line 86, in test_generate_cse
    assert self.generator.subexprs == expected_subexprs
AssertionError

======================================================================
FAIL: pydy.tests.test_models.test_multi_mass_spring_damper_double
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/oliver/repos/math/pydy/pydy/tests/test_models.py", line 71, in test_multi_mass_spring_damper_double
    assert sys.specifieds_symbols == [f0, f1]
AssertionError

----------------------------------------------------------------------
Ran 67 tests in 61.212s
```

But ```pydy.tests.test_models.test_multi_mass_spring_damper_double``` will pass sometimes.

moorepants commented:
I wonder if cse is generating slightly different values for python 2 and 3 and that is causing those remaining failures.

oliverlee commented:
In System, the properties ```constants_symbols``` and ```specifieds_symbols``` are returned as lists but have no inherent order as they are created from set operations.

oliverlee commented:
Ran 67 tests in 65.083s

OK

oliverlee commented:
This now depends on: https://github.com/pydy/pydy/pull/231

oliverlee commented:
Looks like this works with Python 3.3

oliverlee commented:
https://github.com/pydy/pydy/pull/231 must be merged before this one can be.

moorepants commented on README.rst:
```diff
@@ -347,10 +347,11 @@ Release Notes
 
-- Removed the partial implementation of the Mesh shape. [PR `#172`_]
-- Overhauled the code generation package to make the generators more easily
+- Support Python 3.3 and 3.4. [PR `#38`_]
```
Is there a reason for changing all these to present tense?

oliverlee replied:
It's a habit I have from writing git commit messages. I can undo it,

moorepants replied:
If you want to have present tense, that's fine but then they all should be changed. Maybe revert, except for yours and then make a new PR with this and also add a note in `CONTRIBUTING.rst` that says how the release notes should be formatted.

oliverlee replied:
Will revert

moorepants commented on pydy/codegen/code.py:
```diff
@@ -3,28 +3,43 @@
+        pass
+class CythonGeneratorDeprecationWarning(DeprecationWarning):
+        pass
+warnings.simplefilter('once', PyDyCodeGenCodeDeprecationWarning)
```
space here for PEP8

moorepants replied:
And, why are all these special warning classes needed?

oliverlee replied:
Should the pass be 4 spaces or 8 spaces in?

oliverlee replied:
```warnings.simplefilter(once, DeprecationWarning)``` will only show a single DeprecationWarning so if you used multiple soon to be deprecated functions, you'd only get a warning for the first one used.

moorepants replied:
So we can't just have

moorepants replied:
So we can't just have a single `PyDyDepWarning` class? That is what SymPy does. In fact we could just subclass SymPy's dep warning and probably use it as is.

moorepants replied:
As for spaces, I meant blank lines, we typically have 2 spaces between classes and one line other wise.

oliverlee replied:
I realized that but I meant for the indentation for the ```pass``` keyword.

moorepants replied:
Oh, we do 4 spaces.

oliverlee replied:
I'm going to do a PyDyDepWarning that inherits from DeprecationWarning as someone might want to ignore Sympy warnings without know it will also filter PyDy ones.

moorepants replied:
Ok.

moorepants commented on pydy/codegen/tests/test_c_code.py:
```diff
@@ -105,7 +105,17 @@ def test_generate_comma_lists(self):
                     'v0(t), v1(t), v2(t), v3(t), v4(t), v5(t)',
                     'f5(t), f2(t), f3(t), f0(t), f4(t), f1(t)')
```
This line is still the same order as the above list that you ordered.

oliverlee replied:
I just test set equality a few lines down.

moorepants replied:
I see, as per the other discussions I think that switching to sets may not be the best move as the printing code relies on the particular order of the underlying c arrays to be fixed.

moorepants commented on pydy/codegen/tests/test_c_code.py:
```diff
@@ -105,7 +105,17 @@ def test_generate_comma_lists(self):
                     'f5(t), f2(t), f3(t), f0(t), f4(t), f1(t)')
 
-        assert self.generator.comma_lists() == expected
+        actual = self.generator.comma_lists()
```
Does this still test the exact string representation of what is in expected?

oliverlee replied:
Not for the last term. Originally I assumed the order would not be always known, but as the fourth arguments is now a list of the ordered terms, we could.

moorepants replied:
We need to preserve order here because this ultimately is use in the rhs docstrings and headers to tell the user what order the arrays are in.

moorepants commented on pydy/codegen/tests/test_c_code.py:
```diff
@@ -105,7 +105,17 @@ def test_generate_comma_lists(self):
+        assert len(actual) == len(expected)
+        for arg, expected_arg in zip(actual, expected):
+            if expected_arg == expected[-1]:
+                arg_vals = set(x.strip() for x in arg.split(','))
```
The order of the strings is important here, so I'm not sure this does the right thing anymore.

oliverlee replied:
It tests the order for the first 3 args but not the last one.

moorepants commented on pydy/codegen/tests/test_code.py:
```diff
@@ -98,7 +98,7 @@ def test_generate_ode_function(self):
 
         rhs_args = {'constants': constants_map,
-                    'specified': {specifieds[0]: F}}
+                    'specified': {next(iter(specifieds)): F}}
```
I don't understand why this is needed. If specifieds is a list, can't it no longer be index in Py3?

oliverlee replied:
I changed specifieds to a set.

moorepants replied:
I don't like the use of set because it is unordered. I believe all of the code printing relies on the constants, coordinates, speeds, and specifieds to be ordered. It is confusing enough to keep track of them as it is, and guaranteeing order on these public attributes is important.

moorepants replied:
We can leave this, but just to argue some more, :smile_cat:, the previous version of this is much more readable when it was a list.

moorepants commented on pydy/tests/test_models.py:
```diff
@@ -19,8 +16,8 @@ def test_multi_mass_spring_damper():
 
-    assert Counter(sys.constants_symbols) == Counter([k0, c0, m0])
-    assert sys.specifieds_symbols == list()
+    assert sys.constants_symbols == {k0, c0, m0}
```
Does this now return a set? If so, I don't think we should change this, as all the code in the wild depends on this being an ordered sequence.

moorepants replied:
I can see how a set may make more sense for these, but I don't think it should be changed unless you want to properly deprecate it.

oliverlee replied:
What's the procedure for that? I looked through the code and comments and didn't see anything about the list having a specific order. It would have a fixed order once constructed, but it can vary every time you create the the object, as was seen with occasional test failures.

moorepants replied:
The ordering differences are probably due to the fact that various methods in SymPy do not guarantee to spit out the same order of symbols when you search an expression tree. But our System class relies on keeping the order constant once those sequences arrive internally to System.

Which tests were failing wrt to the ordering? We've had some other issues like that too that depend on SymPy versions or even the machine that the the code was run on.

oliverlee replied:
```pydy.tests.test_models.test_multi_mass_spring_damper_double``` would sometimes pass, the order of the variables would change between tests. The other 4 test with failures were failing due to the order changing.

I believe the order is fixed once they arrive to System but I need to look over it again. I'm going to push a new patch set with all changes that you've pointed out _except_ the set thing.

moorepants replied:
Ok, I'll look closer at the set issue.

moorepants commented on pydy/viz/__init__.py:
```diff
@@ -11,25 +11,25 @@
 __all__.extend(light.__all__)
 
-import server
+from . import server
```
Can you add the `__all__.extend(server.__all__)` here just to be complete?

oliverlee replied:
will do

moorepants commented on pydy/viz/tests/test_visualization_frame_scene.py:
```diff
@@ -8,11 +8,11 @@
-from ..camera import PerspectiveCamera, OrthoGraphicCamera
-from ..scene import Scene
-from ..light import PointLight
+from pydy.viz.shapes import Cylinder
```
Do these have to change because of Python 3? If not, I'd prefer to leave them as relative imports.

oliverlee replied:
Python 3 does not allow that.

moorepants replied:
Ok, this is fine then.

moorepants commented:
@oliverlee This is looking awesome! Thanks for fixing this up for Python 3. I have a some comments above that need to be addressed first.

And for the checklist at the top of the PR, we've been making it the responsibility of the reviewer to check those off so far. That way it forces the reviewer to at least think about the unchecked boxes.

moorepants commented on pydy/codegen/code.py:
```diff
@@ -3,28 +3,33 @@
+if parse_version(pydy.__version__) > parse_version('0.4.0'):
+    msg = ("This module, 'pydy.codegen.code', is no longer supported. "
+           "Please remove this module.")
+    raise ValueError(msg.format(v))
```
What is the variable `v`?

oliverlee replied:
whoops doesn't need to be there

moorepants commented on pydy/codegen/ode_function_generators.py:
```diff
@@ -14,6 +16,13 @@
+import pydy
+
+
+class PyDyDepWarning(DeprecationWarning):
```
Maybe we should make this `PyDyDeprecationWarning`, sorry to imply the other earlier.

moorepants replied:
Also, should we put this in `pydy/utils.py`?

oliverlee replied:
Done

oliverlee commented:
How do I cancel jobs in Travis?

moorepants commented:
If you are an owner of the repo you can cancel them. I'm guessing you aren't. I can make you an owner of the pydy org.

moorepants commented:
If you go to the buidl page does it have an "x" in a circle in the top right?

oliverlee commented:
no just a settings dropdown with requests as the only option

moorepants commented:
Try now, I just made you an owner of the PyDy org.

oliverlee commented:
It works now thanks.

oliverlee commented:
I added this commit: https://github.com/oliverlee/pydy/commit/05f769ca176ccd363af2d8636ae5c0f8e224196b
as server.py was missing the definition of __all__.

oliverlee commented:
We need to do something about the minimum versions of module when using Python 3.

moorepants commented on .travis.yml:
```diff
@@ -19,7 +21,12 @@ install:
   - if [[ $DEP_VERSIONS == "oldest" ]]; then
-      conda install numpy==1.7.0 scipy==0.11.0 sympy==0.7.4.1 cython==0.17 theano==0.6.0;
+      conda install numpy==1.8.1 scipy==0.13.3 sympy==0.7.5;
+      if [[ $TRAVIS_PYTHON_VERSION == 2.7 ]]; then
```
Could we do the same thing for numpy, scipy, and sympy? i.e. have two sets of mins, one each for python 2 and 3

This is probably just a matter of building some conda binaries for the older versions with Python 3 support, which I can do at a later date.

moorepants commented:
I've been reading about relative imports vs absolute imports. Relative imports are well supported in Python 2.5+ it seems (including Python 3). They used to be discouraged before a proper implementation was included in Python. This SO question (second answer, in particular) seems to shed light on it:

http://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module

It seems as though it is a subjective preference, with relative imports having a potentially important use case when you may want to rearrange sub-packages easily.

I think that our code base was primarily relative imports before this PR and now it seems to be a mix of the two. I'd prefer to leave the relative imports as is in this PR and continue with that style. But I'm happy if you want to introduce a new PR that changes **all** of the relative imports into absolute imports after this one goes in. This is mostly a subjective matter and all I really care about is consistency. We should choose one pattern and use that all of the time in this codebase. Which pattern we choose can certainly be debated and I don't see much value in debating it. If you feel strongly about it, then I'll accept a separate PR that implements absolute imports across the board.

Is that ok?

moorepants commented:
Another note is that SymPy has tons of circular import issues. It is a real pain the rear sometimes and forces lots of imports to go inside functions/methods. Note in that SO answer that Brandon Rhodes claims that relative imports may prevent that from happening.

moorepants commented on .travis.yml:
```diff
@@ -19,7 +21,11 @@ install:
+        conda install numpy==1.7.0 scipy==0.11.0 sympy==0.7.4.1 cython==0.17 theano==0.6.0;
+      else
+        conda install numpy==1.8.1 scipy==0.13.3 sympy==0.7.5 cython==0.20.1 theano==0.7.0;
+      fi
```
Ok, that looks good. I'll open an issue to build some Python 3 packages for the combination used in 2.7 so that we can have a common min dependencies in the future.

moorepants replied:
Added issue #234.

moorepants commented on README.rst:
```diff
@@ -57,19 +57,19 @@ Installation
 
 PyDy has hard dependencies on the following software:
```
Maybe do this:

PyDy has hard dependencies on the following software\ [#]_:

.. [#] We only test PyDy with these minimum dependencies, but previous versions of NumPy and SciPy are likely to work too.

moorepants commented on pydy/codegen/ode_function_generators.py:
```diff
@@ -417,7 +430,7 @@ def _convert_specifieds_dict_to_array(self, x, t, r):
             if isinstance(type(k), UndefinedFunction):
                 k = (k,)
-            idx = [self.specifieds.index(symmy) for symmy in k]
+            idx = [list(self.specifieds).index(symmy) for symmy in k]
```
Why is this needed? Is `.index` not available on all ordered sequences (lists, tuples)? This allows you to pass in other sequences like sets or custom iterables. I guess that is fine.

moorepants replied:
Note that line 724 still defines `self.inputs` as a list of sequences.

moorepants replied:
The `define_inputs()` method is currently a public method and in the base class it produces a certain type, i.e. list of sequences. But this line is in the subclass. I'd prefer the super method and the method in this class to return the same type.

oliverlee replied:
I changed specifieds to a set.

moorepants replied:
This is the one I have issue with. The user may want their generated rhs function to have the specifieds in a specific order, so if specifieds is a set, how is that maintained?

moorepants replied:
Actually, I don't see where you changed this to a set? It looks like you didn't. Did you?

moorepants replied:
Ok, maybe all this is a misunderstanding because I thought you changed `self.specifieds` to a set here, but I don't see where you actually changed it.

moorepants commented on pydy/codegen/ode_function_generators.py:
```diff
@@ -885,11 +898,11 @@ def __init__(self, *args, **kwargs):
+        specifieds = []
+        if self.specifieds is not None:
+            specifieds = self.specifieds
+        self.inputs = chain(self.coordinates, self.speeds, specifieds,
```
This changes `self.inputs`, a public attribute, from a list to a chain object. I think we need to leave public attributes the same type they were unless you want to deprecate the previous functionality. I'd prefer that to be in a separate PR if we want to deprecate things.

moorepants replied:
Oh yeah, but this code has never been in an official release. So we don't need deprecation. So what is the reason for changing the type of this attribute?

moorepants replied:
Note that line 724 still defines self.inputs as a list of sequences. We should have consistency.

The define_inputs() method is currently a public method and in the base class it produces a certain type, i.e. list of sequences. But this line is in the subclass. I'd prefer the super method and the method in this class to return the same type.

oliverlee replied:
Some of the arguments are lists, some are not. We either have to make them all lists or chain them together. The latter is more efficient.

moorepants replied:
Can you wrap it in `list()` then, as it was before? I don't think efficiency really matters here.

oliverlee replied:
Was this ever public facing? Can we just change self.inputs to an iterator everywhere? Or is there any specific reason it needs to be a list?

moorepants replied:
It was never released, so you can change it to an iterator everywhere.

oliverlee replied:
Also, why is the exact code written twice? It looks to me that it needs only be defined in the abstract base class.

moorepants replied:
I don't know. It looks like that both methods do the same thing. Maybe the one in the Theano generator is a remnant that I forgot to remove. It looks like you can delete it.

moorepants commented on pydy/codegen/tests/test_c_code.py:
```diff
@@ -25,7 +25,7 @@ def setup(self):
                           sys.coordinates,
                           sys.speeds,
-                          sys.specifieds_symbols)
+                          list(sm.ordered(sys.specifieds_symbols)))
```
Yes, this seems correct. The functions in System that extract the constants, coordinates, speeds, and specifieds can extract them in different orders depending on the version of SymPy, the OS, etc. In general, extracting symbols from SymPy expressions does not guarantee some order.

moorepants commented on pydy/codegen/tests/test_code.py:
```diff
@@ -21,7 +21,7 @@ def test_rhs_args(self):
 
         constants = sys.constants_symbols
-        specifieds = sys.specifieds_symbols
+        specifieds = list(sys.specifieds_symbols)
```
Why is this necessary? `sys.specifieds_symbols` is automatically generated and should be a list of SymPy functions (of time).

moorepants commented on setup.py:
```diff
@@ -12,6 +13,18 @@
+    SYMPY_MIN_VER = '0.7.5'
+    CYTHON_MIN_VER = '0.20.1'
+    THEANO_MIN_VER = '0.7.0'
+
```
Thanks for this. What a pain all this dependency management is.

moorepants commented:
Looks like the setup.py file has a small syntax error:

```
$ python setup.py install

  File "setup.py", line 42

    ],

    ^

SyntaxError: invalid syntax
```

moorepants commented:
Looks like theano 0.7.0 requires NumPy 1.9:
```
                          .  0.7.0                np19py34_0  http://repo.continuum.io/pkgs/free/linux-64/ 
                             0.7.0                np19py33_0  http://repo.continuum.io/pkgs/free/linux-64/ 
                          *  0.7.0                np19py27_0  http://repo.continuum.io/pkgs/free/linux-64/ 
                             0.7.0                np19py26_0  http://repo.continuum.io/pkgs/free/linux-64/ 
```

moorepants commented:
I've reviewed this and it is looking really good. I have two issues though:

1. We need to be consistent with relative or absolute imports. My preference is to stay with the relative imports that we had for local packages. But I can see from what I've read that people debate which one is better. So I won't fundamentally hold things up if we choose to do absolute. But whatever we choose we need to be consistent, and I'd really just rather see that change in a separate PR.
2. I am -1 to changing the core attributes of System to unordered sets. After writing tons of code generation code, I think it is much more preferable to have the coordinates, speeds, constants, and specifieds by ordered sequences (I think tuple is the best choice here, not the current lists). @chrisdembia and I designed the `System` class, maybe he can comment here to sway either of us one way or the other. I think the concept of a set for these may be the right abstraction but I think it should be an ordered set (if that exists), otherwise all downstream code will converting these into tuples or lists anyways. Please offer something to convince me otherwise.

moorepants commented:
BTW, we can merge this PR if you revert the two changes I have issue with and we have Python 3 compatibility on master. And then we can open two new PRs for the absolute imports and the changes to System. Would you want to do that? Or discuss here and get it all merged?

moorepants commented:
@sahilshekhawat I'd recommend reviewing this PR as it will be good practice to see these Python 3 changes.

sahilshekhawat commented on setup.py:
```diff
@@ -23,12 +36,13 @@
-                      'scipy>=0.11',
+    install_requires=['numpy>={}'.format(NUMPY_MIN_VER),
+                      'scipy>={}'.format(SCIPY_MIN_VER),
+                      'sympy>={}'.format(SYMPY_MIN_VER,
```
Can you please remove comma from the end?

moorepants replied:
Or is that missing a parentheses?

sahilshekhawat replied:
Yup! a parentheses it is!

moorepants commented:
@oliverlee I really want to get this into the 0.3.0 release. Can you comment ASAP on what you think the solution here is to my concerns? I'm willing to fix them if you don't have time.

I'll think about the set issue some more...maybe that is ok.

oliverlee commented:
I'll take another took at the imports, but I used absolute since I received
errors otherwise. If you see the thread with  Gilbert last year where he
had trouble importing visualizationframe, that was what was happening.

I'll respond to your other issues shortly.
On Jun 16, 2015 8:20 AM, "Jason K. Moore" <notifications@github.com> wrote:

> @oliverlee <https://github.com/oliverlee> I really want to get this into
> the 0.3.0 release. Can you comment ASAP on what you think the solution here
> is to my concerns? I'm willing to fix them if you don't have time.
>
> I'll think about the set issue some more...maybe that is ok.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydy/pydy/pull/229#issuecomment-112307404>.
>

moorepants commented:
I changed some of the imports to test and had no problems. I didn't try them all though.

oliverlee commented:
When is the 0.3.0 release scheduled?

moorepants commented:
The goal was June 15th.

oliverlee commented:
> I am -1 to changing the core attributes of System to unordered sets. After writing tons of code generation code, I think it is much more preferable to have the coordinates, speeds, constants, and specifieds by ordered sequences (I think tuple is the best choice here, not the current lists). @chrisdembia and I designed the System class, maybe he can comment here to sway either of us one way or the other. I think the concept of a set for these may be the right abstraction but I think it should be an ordered set (if that exists), otherwise all downstream code will converting these into tuples or lists anyways. Please offer something to convince me otherwise.

I think a sequence data structure should be avoided unless indexing is important for usage. Preserving the order when iterating can be done if necessary for code generation or testing.

Also ordered sets exist.

moorepants commented:
What is the point of changing these to sets? What does it gain? It worked fined as it was and we get ordered sequences, which I'm claiming are useful. We wrote all this code to work a particular way and I'm not that keen on changing it just for the sake of changing it. Do you have a reasonable argument to why they need to be sets instead of tuples or lists?

oliverlee commented:
We can have ordered sequences without having lists. People might depend on
them being tuples/lists and use index accessing. This should only be
allowed if intended by design.

We also use a number of set operations on these fields, requiring changing
between lists and sets.


On Tue, Jun 16, 2015 at 10:11 AM, Jason K. Moore <notifications@github.com>
wrote:

> What is the point of changing these to sets? What does it gain? It worked
> fined as it was and we get ordered sequences, which I'm claiming are
> useful. We wrote all this code to work a particular way and I'm not that
> keen on changing it just for the sake of changing it. Do you have a
> reasonable argument to why they need to be sets instead of tuples or lists?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydy/pydy/pull/229#issuecomment-112327400>.
>

moorepants commented:
Ok, I'm fine with changing to sets for `constants_symbols` and `specifieds_symbols` but I don't see any reason that denying people index access to them is a bad thing.

I just heard 5 gunshots outside my window, btw...

oliverlee commented:
Well if this was released before and people had depended on index access,
then it would be much harder to change these fields from lists to sets.  ;)

Are you in Oakland?
On Jun 16, 2015 10:24 AM, "Jason K. Moore" <notifications@github.com> wrote:

> Ok, I'm fine with changing to sets for constants_symbols and
> specifieds_symbols but I don't see any reason that denying people index
> access to them is a bad thing.
>
> I just heard 5 gunshots outside my window, btw...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydy/pydy/pull/229#issuecomment-112331014>.
>

moorepants commented:
To be clear I'm fine with `System.constants_symbols` and `System.specifieds_symbols` being sets. Is that even what you've changed?

But I think you changed `ODEFunctionGenerator.specifieds` to a set too. That one I am not sure about. My intention in the generator designs were to force the users to pass in ordered sequences, sot that they have the power to control the order of the array arguments to the generated rhs. I'll look at your changes again for that one.

moorepants commented:
I'm in Oak Park, Sacramento...

oliverlee commented:
> To be clear I'm fine with System.constants_symbols and System.specifieds_symbols being sets. Is that even what you've changed?
Yes

> But I think you changed ODEFunctionGenerator.specifieds to a set too. 
I'll take a look. I didn't realize there was another ```specifieds``` field.

moorepants commented:
Are you on giitter, it is easier to chat there?

moorepants commented:
https://gitter.im/pydy/pydy

oliverlee commented:
Had increase SciPy to 0.14.0 after changing NumPy to 1.9.
Relative import commit was deleted and I changed https://github.com/oliverlee/pydy/commit/10bd4cdb350f57db85ea06b5f924416afcf72944 to use relative imports as well.

I think this has now addressed all of your points.

moorepants commented:
Thanks for this heroic PR! Merging!
